### ISSUE TITLE
[5.1 beta] Working with nested foreach Describe > Context/It

### ISSUE THREAD

**Issue by wsmelton** [pester/Pester#1731]
## 1. General summary of the issue

Writing tests using 5.1 (which is currently beta/pre-release) and I'm using `Describe -ForEach` with no issues, but having trouble when I need to nest another `-ForEach` in a Context block or even just an `It -TestCases`.

## 2. Describe Your Environment

Operating System, Pester version, and PowerShell version:

```powershell
Pester version     : 5.1.0 
PowerShell version : 5.1.19041.546
OS version         : Microsoft Windows NT 10.0.19041.0
```

Using VS Code Insiders with Windows PowerShell 5.1.

I'm trying to test the help and want to get detailed in this manner. So I'm looping over the commands for the Describe block. In a Context or It block though I want to iterate over a nested object and not able to get that working.

Tests I've written so far:

```powershell
$ModuleRoot = (Resolve-Path "$PSScriptRoot\..").Path
$ModuleName = (Get-ChildItem "$ModuleRoot\*.psd1").Name.Trim(".psd1")
$CommandPath = "$ModuleRoot\functions"

$includedNames = (Get-ChildItem $CommandPath -Recurse -File | Where-Object Name -like "*.ps1").BaseName
$commands = Get-Command -Module $ModuleName -CommandType Function | Where-Object Name -In $includedNames

Describe "Test help for module" -ForEach $commands {
    BeforeAll {
        $commandName = $_.Name
        $help = Get-Help $commandName -ErrorAction SilentlyContinue
        $examples = $help.Examples.example
    }
    Context "Test <_> for basic help" {
        It "Should not be auto-generated help" {
            $help.Synopsis | Should -Not -BeLike "*<CommonParameters>*"
        }
        It "Should have a Description" {
            $help.Description | Should -Not -BeNullOrEmpty
        }
        It "Should have an Example" {
            $help.Examples | Select-Object -First 1 | Should -HaveCount 1
        }
        It "Each Example should have an introduction of PS C:\>" -TestCases $examples {
            $_.Introduction.Text | Should -Be "PS C:\>"
        }
    }
}
```
## 3. Expected Behavior

I would at least expect that last It block to execute but it is not even being picked up in the discovery for some reason.

Diagnostic output:

```
PS 5.1] [21:39:07] [  7.8 s ][ git:master ] C:\git\thycotic.secretserver
 [17] > Invoke-Pester .\tests\Module.Help.Tests.ps1 -Output Diagnostic
Discovery: Starting test discovery in 1 test containers. 

Starting discovery in 1 files.
Discovery: Discovering tests in C:\git\thycotic.secretserver\tests\Module.Help.Tests.ps1 
Discovering in C:\git\thycotic.secretserver\tests\Module.Help.Tests.ps1.
Found 27 tests. 171ms
Discovery: Found 27 tests in 185 ms 
Discovery: Processing discovery result objects, to set root, parents, filters etc. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help) Block will be included in the run, because there were no include filters, and will let its children to determine whether or not it should run. 
Filter: (Test help for module.Test <_> for basic help.Should not be auto-generated help) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have a Description) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Filter: (Test help for module.Test <_> for basic help.Should have an Example) Test will be included in the run, because there were no include filters so all tests are included unless they match exclude rule. 
Discovery finished in 1.55s.
Discovery: Test discovery finished. 

Running tests from 'C:\git\thycotic.secretserver\tests\Module.Help.Tests.ps1'
 Context Test Disable-TssSecret for basic help
   [+] Should not be auto-generated help 28ms (16ms|12ms)
   [+] Should have a Description 22ms (15ms|7ms)
   [+] Should have an Example 24ms (15ms|8ms)
 Context Test Get-Overit for basic help
   [+] Should not be auto-generated help 20ms (10ms|10ms)
   [+] Should have a Description 15ms (7ms|8ms)
   [+] Should have an Example 15ms (7ms|8ms)
 Context Test Get-TssSecret for basic help
   [+] Should not be auto-generated help 20ms (8ms|11ms)
   [+] Should have a Description 15ms (9ms|6ms)
   [+] Should have an Example 15ms (8ms|6ms)
 Context Test Get-TssSecretTemplate for basic help
   [+] Should not be auto-generated help 21ms (9ms|12ms)
   [+] Should have a Description 14ms (8ms|6ms)
   [+] Should have an Example 22ms (11ms|11ms)
 Context Test Get-TssSession for basic help
   [+] Should not be auto-generated help 23ms (13ms|10ms)
   [+] Should have a Description 21ms (11ms|10ms)
   [+] Should have an Example 18ms (12ms|6ms)
 Context Test Invoke-TssRestApi for basic help
   [+] Should not be auto-generated help 21ms (13ms|9ms)
   [+] Should have a Description 18ms (11ms|8ms)
   [+] Should have an Example 22ms (13ms|10ms)
 Context Test New-TssSession for basic help
   [+] Should not be auto-generated help 20ms (8ms|12ms)
   [+] Should have a Description 16ms (9ms|7ms)
   [+] Should have an Example 20ms (14ms|6ms)
 Context Test Search-TssSecret for basic help
   [+] Should not be auto-generated help 26ms (14ms|12ms)
   [+] Should have a Description 21ms (8ms|13ms)
   [+] Should have an Example 17ms (9ms|8ms)
 Context Test Set-TssSecret for basic help
   [+] Should not be auto-generated help 30ms (17ms|13ms)
   [+] Should have a Description 24ms (15ms|8ms)
   [+] Should have an Example 22ms (13ms|10ms)
Tests completed in 7.82s
Tests Passed: 27, Failed: 0, Skipped: 0 NotRun: 0
```

**Comment by nohwnd** [comment]
You define the examples in BeforeAll, and that won't run till Run, so during discovery there are no examples and no It is generated. I will add a check and make it throw by default if the collection is empty.

**Comment by nohwnd** [comment]
Took a stab on this, and it is a bit annoying to just throw during discovery because it will fail the whole run when there is any mistake. A better way would be to emit a failing test that would be added into the run.

**Comment by wsmelton** [comment]
Well, part of what I do not see is the `$examples` is not null, it does have content. The first function I hit has one example and debugging shows the TestCases object is populated.

![image](https://user-images.githubusercontent.com/11204251/97085218-b1d0ad00-15e1-11eb-82ac-df4aba1c48ff.png)

**Comment by wsmelton** [comment]
If I put breakpoints on line 15 and line 26:
![image](https://user-images.githubusercontent.com/11204251/97086373-738abc00-15e8-11eb-9d47-e880bca8ecea.png)

When I start debugging I guess that the discovery process runs first and the breakpoint hits on line 26, never on line 15 until the test starts running.

To get around this should I be doing this differently or is this expected behavior for this type of use case?

**Comment by nohwnd** [comment]
This is how it works, BP on 26 is hit first because during discovery all the Describe, Context, It and BeforeAll functions will run, but `It` and `BeforeAll` won't invoke the script block that is associated with it. The It that uses -TestCases in $examples runs and $examples is null.

Then during Run phase the scriptblocks from each BeforeAll and It will run. And that is when $examples get defined. But at that point in time it's too late.

Passing the values around like this should work. That will run the setup during discovery and will associate the $help with the context so you can use it later in your tests.

```powershell
BeforeDiscovery {
    $ModuleRoot = (Resolve-Path "$PSScriptRoot\..").Path
    $ModuleName = (Get-ChildItem "$ModuleRoot\*.psd1").Name.Trim(".psd1")
    $CommandPath = "$ModuleRoot\functions"

    $includedNames = (Get-ChildItem $CommandPath -Recurse -File | Where-Object Name -like "*.ps1").BaseName
    $commands = Get-Command -Module $ModuleName -CommandType Function | Where-Object Name -In $includedNames
}

Describe "Test help for module" -ForEach $commands {
    BeforeDiscovery {
        $help = Get-Help $_.Name -ErrorAction SilentlyContinue
        $examples = $help.Examples.example
    }

    Context "Test <_> for basic help" -ForEach @{ Help = $help } {
        It "Should not be auto-generated help" {
            $help.Synopsis | Should -Not -BeLike "*<CommonParameters>*"
        }
        It "Should have a Description" {
            $help.Description | Should -Not -BeNullOrEmpty
        }
        It "Should have an Example" {
            $help.Examples | Select-Object -First 1 | Should -HaveCount 1
        }
        It "Each Example should have an introduction of PS C:\>" -TestCases $examples {
            $_.Introduction.Text | Should -Be "PS C:\>"
        }
    }
}
```

**Comment by DEberhardt** [comment]
This is helpful, thank you.
I too, tried with BeforeAll and had no luck. Is there a document somewhere describing the Parsing order you outlined, or should that be intuitive?

**Comment by wsmelton** [comment]
So use of `BeforeDiscovery` in the `Describe` and `Context` causes more test to fail compared to using `BeforeAll` in Context block. The BeforeDiscovery is at least failing on the test but should be passing on most of my functions because they pass the test. 

Note, I've updated to the latest beta release as well: `5.1.0-beta3`

With BeforeAll in Context:

![image](https://user-images.githubusercontent.com/11204251/97113664-e3b54280-16b9-11eb-9a89-5220fa851214.png)

vs BeforeDiscovery:

![image](https://user-images.githubusercontent.com/11204251/97113686-047d9800-16ba-11eb-9205-2f11400bd69c.png)

**Comment by nohwnd** [comment]
@wsmelton  you did not pass the $help to the Context. Looking back I should have pointed it out more visibly.

```powershell
    Context "Test <_> for basic help" -ForEach @{ Help = $help } {
```

**Comment by nohwnd** [comment]
@DEberhardt I don't think there is document that is specific to this. https://github.com/pester/pester#discovery--run this is how I described it for the readme.

**Comment by wsmelton** [comment]
Is there a reason the Context block requires different syntax than the Describe block?

**Comment by wsmelton** [comment]
Adding the hashtable to Context seems to affect the output of the text for the Context block. The reference to `<_>` is being caught now and replaced with `System.Collections.Hashtable` instead of the command from Describe block, is that intended? If so, do I need to just set a variable to keep the command name there in the text?

![image](https://user-images.githubusercontent.com/11204251/97172296-d69d5f80-175c-11eb-9971-07701ca3b6c8.png)

**Comment by wsmelton** [comment]
I changed it to this and that seems to work for what I need... much-appreciated @nohwnd (I think once the doc site is updated with information on these it will help users adopt it more easily.)

```powershell
BeforeDiscovery {
    $ModuleRoot = (Resolve-Path "$PSScriptRoot\..").Path
    $ModuleName = (Get-ChildItem "$ModuleRoot\*.psd1").Name.Trim(".psd1")
    $CommandPath = "$ModuleRoot\functions"

    $includedNames = (Get-ChildItem $CommandPath -Recurse -File | Where-Object Name -like "*.ps1").BaseName
    $commands = Get-Command -Module $ModuleName -CommandType Function | Where-Object Name -In $includedNames
}
Describe "Test help for module" -ForEach $commands {
    BeforeDiscovery {
        $commandName = $_.Name
        $help = Get-Help $_.Name
        $examples = $help.Examples.example
    }
    Context "Test $commandName for basic help" -Foreach @{help = $help} {
        It "Should not be auto-generated help" {
            $help.Synopsis | Should -Not -BeLike "*<CommonParameters>*"
        }
        It "Should have a Synopsis" {
            $help.Synopsis | Should -Not -BeNullOrEmpty
        }
        It "Should have a Description" {
            $help.Description | Should -Not -BeNullOrEmpty
        }
        It "Should have an Example" {
            $help.examples | Select-Object -First 1 | Should -HaveCount 1
        }
        It "Each Example should have an introduction of PS C:\>" -TestCases $examples {
            $_.introduction.Text | Should -Be "PS C:\>"
        }
    }
}
```

![image](https://user-images.githubusercontent.com/11204251/97173627-de5e0380-175e-11eb-8776-f9bfa5a39461.png)

**Comment by simonsabin** [comment]
I think pester should be stricter in this regard to force use of blocks for scripts. It would force people to realise the different execution phases of Disovery and Run. Raised the following as a suggestion #1763

**Comment by fflaten** [comment]
Fixed in pester/docs#70 ?
https://pester.dev/docs/usage/data-driven-tests

**Comment by DarkLite1** [comment]
This tricked me too. I had some `TestCases` in a `BeforeAll` that where not triggering any tests. It appears that `BeforeDiscovery` fixes this. Thanks for Providing this info.

Small repro:

```powershell
Describe 'Show-MenuHC' {
    $testCases = @(1 , 2)
    Context 'triggered with testCases 1, 2 not 3, 5' {
        BeforeAll {
            $testCases = @(3 , 4)
        }
        Describe '<_>' -Foreach $testCases {
            BeforeAll {
              
            }
            It 'green' {
                1 | Should -Be 1
            }
        }
    }
    Context 'Triggered because testCases are in BeforeDiscovery' {
        BeforeDiscovery {
            $testCases = @(5 , 6)
        }
        Describe '<_>' -Foreach $testCases {
            It 'orange' {
                1 | Should -Be 1
            }
        }
    } 
}
```
